Signal/src/contact/SystemContactsFetcher.swift

```diff
@@ -0,0 +1,120 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import Contacts
+
+@objc protocol SystemContactsFetcherDelegate: class {
+    func systemContactsFetcher(_ systemContactsFetcher: SystemContactsFetcher, updatedContacts contacts: [Contact])
+}
+
+@objc
+class SystemContactsFetcher: NSObject {
+
+    private let TAG = "[SystemContactsFetcher]"
+
+    public weak var delegate: SystemContactsFetcherDelegate?
+
+    public var authorizationStatus: CNAuthorizationStatus {
+        return CNContactStore.authorizationStatus(for: CNEntityType.contacts)
+    }
+
+    public var isAuthorized: Bool {
+        guard self.authorizationStatus != .notDetermined else {
+            assertionFailure("should have called `requestOnce` before this point.")
+            Logger.error("\(TAG) should have called `requestOnce` before checking authorization status.")
+            return false
+        }
+
+        return self.authorizationStatus == .authorized
+    }
+
+    private let contactStore = CNContactStore()
+    private var systemContactsHaveBeenRequestedAtLeastOnce = false
```

Comment:
nope, it's just a debounce. 

Disregarding updates, we only want to pull in the address book one time. However this could happen at various points in the app. So at each of these checkpoints, we fetch the address book only if it hasn't been previously fetched.

(And then we also start fetching updates at this point with the observer notification)